src/iosrv/drpc_client.c

```diff
@@ -158,6 +161,165 @@ notify_bio_error(int media_err_type, int tgt_id)
+			 tm->tm_hour, tm->tm_min, tm->tm_sec,
+			 (long int)tv.tv_usec / 10000);
+	}
+
+	/** XXX: nodename should be fetched only once and not on every call */
+	D_ALLOC(evt->hostname, DAOS_RAS_EVENT_STR_MAX_LEN);
+	if (evt->hostname == NULL) {
+		D_ERROR("failed to allocate hostname\n");
+		D_FREE(evt->timestamp);
+		return -DER_NOMEM;
+	}
+	(void)uname(&uts);
+	D_STRNDUP(evt->hostname, uts.nodename, DAOS_RAS_EVENT_STR_MAX_LEN - 1);
+
+	/* TODO: add rank id to event */
+
+	D_DEBUG(DB_MGMT, "&&& RAS EVENT id: [%s] ts: [%s] host: [%s] sev: [%s]"
+		" type: [%s] puuid: ["DF_UUID"] msg: [%s]",
+		evt->id, evt->timestamp, evt->hostname,
+		ras_event_sev2str(evt->severity), ras_event_type2str(evt->type),
+		DP_UUID(evt->pool_svc_info->pool_uuid), evt->msg);
+
+	return 0;
+}
+
+static void
+free_ras(Mgmt__RASEvent *evt)
+{
+	D_FREE(evt->hostname);
+	D_FREE(evt->timestamp);
+}
+
+static int
+notify_ras_event(Mgmt__RASEvent *evt)
+{
+	Mgmt__ClusterEventReq	 req = MGMT__CLUSTER_EVENT_REQ__INIT;
+	uint8_t			*reqb;
+	size_t			 reqb_size;
+	Drpc__Call		*dreq;
+	Drpc__Response		*dresp;
+	int			 rc;
+
+	if (evt == NULL) {
+		D_ERROR("invalid RAS event\n");
+		return -DER_INVAL;
+	}
+
+	req.event_case = MGMT__CLUSTER_EVENT_REQ__EVENT_RAS;
+	req.ras = evt;
+
+	reqb_size = mgmt__cluster_event_req__get_packed_size(&req);
+	D_ALLOC(reqb, reqb_size);
+	if (reqb == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	mgmt__cluster_event_req__pack(&req, reqb);
+	rc = drpc_call_create(dss_drpc_ctx, DRPC_MODULE_MGMT,
+			      DRPC_METHOD_MGMT_CLUSTER_EVENT, &dreq);
+	if (rc != 0) {
+		D_FREE(reqb);
+		goto out;
+	}
+
+	dreq->body.len = reqb_size;
+	dreq->body.data = reqb;
+
+	rc = drpc_call(dss_drpc_ctx, R_SYNC, dreq, &dresp);
+	if (rc != 0)
+		goto out_dreq;
+	if (dresp->status != DRPC__STATUS__SUCCESS) {
+		D_ERROR("received erroneous dRPC response: %d\n",
+			dresp->status);
+		rc = -DER_IO;
+	}
+
+	drpc_response_free(dresp);
+out_dreq:
+	/* This also frees reqb via dreq->body.data. */
+	drpc_call_free(dreq);
+out:
+	return rc;
+}
+
+int
+ds_notify_pool_svc_update(uuid_t pool_uuid, d_rank_list_t *svc)
+{
+	Mgmt__RASEvent		 evt = MGMT__RASEVENT__INIT;
+	Mgmt__PoolSvcEventInfo	 info = MGMT__POOL_SVC_EVENT_INFO__INIT;
+	int			 rc;
+
+	if (dss_drpc_ctx == NULL) {
+		D_ERROR("DRPC not connected\n");
+		return -DER_UNINIT;
+	}
+
+	D_ASSERT(pool_uuid != NULL);
+
+	D_ALLOC(info.pool_uuid, DAOS_UUID_STR_SIZE);
+	if (info.pool_uuid == NULL) {
+		D_ERROR("failed to allocate pool uuid\n");
+		return -DER_NOMEM;
+	}
+	uuid_unparse_lower(pool_uuid, info.pool_uuid);
+
+	D_ASSERT(svc != NULL && svc->rl_nr > 0);
+
+	rc = rank_list_to_uint32_array(svc, &info.svc_reps, &info.n_svc_reps);
+	if (rc != 0) {
+		D_ERROR("failed to convert svc replicas to proto\n");
+		goto out_uuid;
+	}
+
+	evt.extended_info_case = MGMT__RASEVENT__EXTENDED_INFO_POOL_SVC_INFO;
+	evt.pool_svc_info = &info;
+
+	rc = init_ras("pool_svc_ranks_update", (uint32_t)RAS_SEV_INFO,
```

Comment:
Hmm. This doesn't seem correct. I understand that we wanted to move toward having the event IDs defined as a 64-bit string, but I would have expected that we define them as constants on the C side. Then we can access them from both languages and they're maintained in one place.

e.g.
```C
#define RAS_POOL_SVC_RANKS_UPDATE "pool_svc_ranks_update"
```

@johannlombardi: I haven't had a chance to look at your proposal yet, but it seems like Tom has incorporated some of those ideas here. What's your take on where/how these IDs should be defined?

Reply:
You and @johannlombardi seem to be asking for opposite approaches, my original approach was to define in ras.h but Johann specifically asked to only be defined at the call to raise the event.

Reply:
OK, so we should regroup with @johannlombardi on this. From my perspective, just using bare strings at the site of raising an event is not ideal, because it means that we have to guarantee that all references to that event id use the same string. I figured we would have a single list of event IDs set up as `#define`s in a header, and this list would be accessible from both languages.

Reply:
that was my initial approach